src/radical/pilot/session.py

```diff
@@ -62,379 +54,174 @@ def __init__(self, dburl=None, uid=None, cfg=None, _connect=True):
-
-        # The resource configuration dictionary associated with the session.
-        self._resource_configs = {}
-
-        # if a config is given, us its values:
-        if cfg:
-            self._cfg = copy.deepcopy(cfg)
-        else:
-            # otherwise we need a config
-            self._cfg = ru.read_json("%s/configs/session_%s.json"
-                    % (os.path.dirname(__file__),
-                       os.environ.get('RADICAL_PILOT_SESSION_CFG', 'default')))
-
-        # fall back to config data where possible
-        # sanity check on parameters
-        if not uid :
-            uid = self._cfg.get('session_id')
-
-        if uid:
-            self._uid         = uid
-            self._reconnected = True
-        else:
-            # generate new uid, reset all other ID counters
-            # FIXME: this will screw up counters for *concurrent* sessions,
-            #        as the ID generation is managed in a process singleton.
-            self._uid = ru.generate_id('rp.session',  mode=ru.ID_PRIVATE)
-            ru.reset_id_counters(prefix='rp.session', reset_all_others=True)
-
-        if not self._cfg.get('session_id'): self._cfg['session_id'] = self._uid
-        if not self._cfg.get('owner')     : self._cfg['owner']      = self._uid
-        if not self._cfg.get('logdir')    : self._cfg['logdir']     = '%s/%s' \
-                                                     % (os.getcwd(), self._uid)
-        self._logdir = self._cfg['logdir']
-        self._prof   = self._get_profiler(name=self._cfg['owner'])
-        self._rep    = self._get_reporter(name=self._cfg['owner'])
-        self._log    = self._get_logger  (name=self._cfg['owner'],
-                                          level=self._cfg.get('debug'))
-
-        if _connect:
+            * **cfg** (`str` or `dict`): a named or instantiated configuration
+              to be used for the session.
+
+            * **uid** (`string`): Create a session with this UID.  Session UIDs
+              MUST be unique - otherwise they will lead to conflicts in the
+              underlying database, resulting in undefined behaviours (or worse).
+
+            * **_primary** (`bool`): only sessions created by the original
+              application process (via `rp.Session()`, will connect to the  DB.
+              Secondary session instances are instantiated internally in
+              processes spawned (directly or indirectly) by the initial session,
+              for example in some of it's components.  A secondary session will
+              inherit the original session ID, but will not attempt to create
+              a new DB collection - if such a DB connection is needed, the
+              component needs to establish that on its own.
+        '''
 
-            # we need a dburl to connect to.
-            if not dburl:
-                dburl = os.environ.get("RADICAL_PILOT_DBURL")
+        # NOTE: `name` and `cfg` are overloaded, the user cannot point to
```

Comment:
I would add the following to the previous docstring so that it goes into our documentation automatically.

Reply:
I don't think so: the docstring for the `cfg` parameter already mentions the overloading (``` (`str` or `dict`): a named or instantiated configuration```) - I only pointed it out in an extra comment as that is one of the very view places in the API where we overload based on type.